### PR TITLE
[WIP] Allow for scripts to be tested

### PR DESCRIPTION
**Pull request by clayellis**
This is a work in progress PR for Issue #77. Any input up to this point would be appreciated! There are a few obvious blocks of code that need to be moved, (default file contents for LinuxMain.swift), and a couple comments that will be removed in the final PR.

- [x] Able to create script & tests using `marathon create MyNewScript --tests` 
- [x] Able to create tests (for existing script) using `marathon create MyScript --tests`
- [x] Able to edit tests using `marathon edit MyScript --tests`
- [ ] Able to run tests using `marathon test MyScript`

### CODE REVIEW

File: Sources/MarathonCore/Create.swift [hunk 1]
```diff
@@ -38,25 +38,74 @@ extension CreateError: PrintableError {
...
+                try createScriptTests(for: script)
+                try beginEditing(script, if: shouldOpen)
+            } else {
+                let file = try createScript(at: path)
+                let script = try scriptManager.script(at: file.path)
+                try createScriptTests(for: script)
+                try beginEditing(script, if: shouldOpen)
+            }
+        } else {
+            let file = try createScript(at: path)
```
Comment (JohnSundell) [inline Sources/MarathonCore/Create.swift]: Hmmm there should be a way to reduce these code paths. What we want is basically:

1. Create the script file
2. Create the tests if the `--tests` flag has been passed.

Shouldn't this be able to be expressed like this:

```swift
let script = try createScript()

if arguments.contains("--tests") {
    try createTests()
}
```

Maybe I'm over-simplifying, but basically looking for one of the operations to just run on top of the other, instead of having multiple nested `if/else` statements. What do you think @clayellis?

File: Sources/MarathonCore/ScriptManager.swift [hunk 2]
```diff
@@ -133,6 +137,14 @@ internal final class ScriptManager {
             try removeDataForScript(at: path)
         }
     }
+    
+    func addTest(file testFile: File, to script: Script) throws {
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: Maybe `addTestFile(...` since "Test file" is the noun we're referring to?

File: Sources/MarathonCore/ScriptManager.swift [hunk 3]
```diff
@@ -242,6 +254,32 @@ internal final class ScriptManager {
 
         return scriptFolder
     }
+    
+    @discardableResult private func createTestsFolderIfNeeded(for script: Script, file: File) throws -> Folder {
+        if (try? script.folder.file(named: "OriginalTestsFile")) == nil {
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: You can use `script.folder.containsFile(named: "OriginalTestsFile")` here.

File: Sources/MarathonCore/ScriptManager.swift [hunk 4]
```diff
@@ -242,6 +254,32 @@ internal final class ScriptManager {
...
+            try script.folder.createSymlink(to: file.path, at: "OriginalTestsFile", printer: printer)
+        }
+        
+        let testsFolder = try script.folder.createSubfolderIfNeeded(withName: "Tests")
+        try testsFolder.empty()
+        
+        let testsModuleFolder = try testsFolder.createSubfolder(named: script.name + "Tests")
+        try testsModuleFolder.empty()
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: You don't need to empty this folder, it was just created 🙂

File: Sources/MarathonCore/ScriptManager.swift [hunk 5]
```diff
@@ -242,6 +254,32 @@ internal final class ScriptManager {
...
+        try testsModuleFolder.createFile(named: file.name, contents: file.read())
+        
+        // TODO: Determine why LinuxMain.swift is missing from the .xcodeproj when opened.
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: Yeah it's never part of the Xcode project, so you're assumption is correct 👍

File: Sources/MarathonCore/Create.swift [hunk 6]
```diff
@@ -38,25 +38,74 @@ extension CreateError: PrintableError {
...
         guard let path = arguments.first?.asScriptPath() else {
             throw Error.missingName
         }
-
+        
+        if arguments.contains("--tests") {
+            if let scriptFile = try? FileSystem().currentFolder.file(atPath: path),
+                let script = try? scriptManager.script(at: scriptFile.path) {
+                try createScriptTests(for: script)
+                try beginEditing(script, if: shouldOpen)
```
Comment (clayellis) [inline Sources/MarathonCore/Create.swift]: I agree. I need to clean this section up a bit.

File: Sources/MarathonCore/Create.swift (hunk 1 above)
Comment (clayellis) [inline Sources/MarathonCore/Create.swift]: There are three combinations of actions we can take: 
1. Create a new script without tests
2. Create a new script and tests together
3. Create tests for an existing script (without overwriting the script)

Hence the three branches. Some of the surrounding logic can definitely be simplified, though.

File: Sources/MarathonCore/ScriptManager.swift (hunk 2 above)
Comment (clayellis) [inline Sources/MarathonCore/ScriptManager.swift]: Agreed!

File: Sources/MarathonCore/ScriptManager.swift (hunk 3 above)
Comment (clayellis) [inline Sources/MarathonCore/ScriptManager.swift]: Awesome—should I update the syntax in `createFolderIfNeededForScript(...)` as well? Or leave that for another PR?

File: Sources/MarathonCore/ScriptManager.swift (hunk 4 above)
Comment (clayellis) [inline Sources/MarathonCore/ScriptManager.swift]: I think I meant to use `createSubfolderIfNeeded`.

File: Sources/MarathonCore/Create.swift [hunk 7]
```diff
@@ -38,25 +38,74 @@ extension CreateError: PrintableError {
...
+    }
+    
     func execute() throws {
         guard let path = arguments.first?.asScriptPath() else {
             throw Error.missingName
         }
-
+        
+        if arguments.contains("--tests") {
+            if let scriptFile = try? FileSystem().currentFolder.file(atPath: path),
```
Comment (clayellis) [inline Sources/MarathonCore/Create.swift]: In order to avoid overwriting it, I think so, yeah.

**Comment by clayellis** [comment]
@JohnSundell I made a few changes based on your input—let me know what you think! I'll start working on `edit --tests` later tonight.

**Comment by clayellis** [comment]
@JohnSundell I've left a few comments for you to check in `Script.swift` concerning the implementation for `edit --tests`.

File: Sources/MarathonCore/Create.swift [hunk 8]
```diff
@@ -38,12 +38,12 @@ extension CreateError: PrintableError {
 
 internal final class CreateTask: Task, Executable {
     private typealias Error = CreateError
-
+
```
Comment (JohnSundell) [inline Sources/MarathonCore/Create.swift]: Please unstage these changes 😉 Pro tip is to make Xcode automatically remove this kind of whitespace for you:
screen shot 2017-06-04 at 18 38 59

File: Sources/MarathonCore/Create.swift [hunk 9]
```diff
@@ -55,20 +55,55 @@ internal final class CreateTask: Task, Executable {
...
+        }
 
+        if !argumentsContainNoOpenFlag {
+            let script = try scriptManager.script(at: scriptFile.path, allowRemote: false)
+            try script.edit(arguments: arguments, open: true)
+        }
+    }
+    
+    private func createScriptIfNeeded(atPath path: String) throws -> File {
+        if arguments.contains("--tests"), let file = try? Folder.current.file(atPath: path) {
```
Comment (JohnSundell) [inline Sources/MarathonCore/Create.swift]: Why is `arguments.contains("--tests")` a condition here? Don't we always want to return if the file already exists?

File: Sources/MarathonCore/Create.swift [hunk 10]
```diff
@@ -55,20 +55,55 @@ internal final class CreateTask: Task, Executable {
...
         printer.output("🐣  Created script at \(path)")
 
-        if !argumentsContainNoOpenFlag {
-            let script = try scriptManager.script(at: file.path, allowRemote: false)
-            try script.edit(arguments: arguments, open: true)
+        return file
+    }
+    
+    @discardableResult private func createTests(atPath path: String) throws -> File {
+        let script = try scriptManager.script(at: path, allowRemote: false)
```
Comment (JohnSundell) [inline Sources/MarathonCore/Create.swift]: Since you already have the script when you are calling this method, how about taking it as an argument instead?

File: Sources/MarathonCore/Script.swift [hunk 11]
```diff
@@ -131,6 +136,35 @@ internal final class Script {
...
+                // open the second file in a new window. By sleeping for a second, we give Xcode a reasonable chance to finish opening
+                // the first file before trying to open the second. By sleeping at this point, the user shouldn't notice a difference
+                // except for the "i" message below about commits, since the first open has already been executed.
+                //
+                // Other approaches I've tried for avoiding potentially opening a second window:
+                // 1. Edit the .xcuserstate before opening Xcode to spoof the window state and force Xcode to open to a specific file
+                //
+                // I've asked a question on StackOverflow to try find a better way to do this:
+                // https://stackoverflow.com/questions/44346442/open-xcode-project-and-set-active-file-from-terminal
+                sleep(1)
```
Comment (JohnSundell) [inline Sources/MarathonCore/Script.swift]: This is clever! 😄 I've been experimenting with different ways to get Xcode to open to a specific file, both for Marathon (would be super nice if it opened at the script file directly), and for tools like TestDrive. I never considered that Xcode actually auto-opens the file if a project containing it is already open, so this is super awesome 🚀 I'm thinking, instead of sleeping, how about we observe whether Xcode is opened? Should be a way to do that using the system APIs.

File: Sources/MarathonCore/ScriptManager.swift [hunk 12]
```diff
@@ -259,6 +271,25 @@ internal final class ScriptManager {
...
+    @discardableResult private func createTestsFolderIfNeeded(for script: Script, file: File) throws -> Folder {
+        if !script.folder.containsFile(named: "OriginalTestsFile") {
+            try script.folder.createSymlink(to: file.path, at: "OriginalTestsFile", printer: printer)
+        }
+        
+        let testsFolder = try script.folder.createSubfolderIfNeeded(withName: "Tests")
+        try testsFolder.empty()
+        
+        let testsModuleFolder = try testsFolder.createSubfolderIfNeeded(withName: script.name + "Tests")
+        try testsModuleFolder.empty()
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: No need to empty here, you are already emptying the parent folder on line `281`.

File: Sources/MarathonCore/ScriptManager.swift [hunk 13]
```diff
@@ -321,4 +352,14 @@ internal final class ScriptManager {
             return path
         }
     }
+    
+    private func makeLinuxMainTestFileContent(for script: Script) -> String {
+        let content = "import XCTest\n" +
```
Comment (JohnSundell) [inline Sources/MarathonCore/ScriptManager.swift]: How about just returning the expression directly instead of having this temporary variable?

**Comment by JohnSundell** [comment]
@clayellis Any updates on this one? 🙂 Let me know if you need any help!

**Comment by clayellis** [comment]
@JohnSundell I apologize! My wife and I just had a baby and I've had next to no time to put towards open source! Things are starting to smooth out now so I'll get back to contributing. I'm sorry for leaving you in the dark on this one!

**Comment by JohnSundell** [comment]
Going to close this one, sorry @clayellis. I'm limiting the scope of Marathon going forward, and adding support for tests won't make the cut. Thanks anyway for all of your work on this, I really appreciate it.

**Comment by clayellis** [comment]
@johnsundell no worries! I'm all for a tight scope 🙂
